WPIcal: Field Calibration Tool

ElliotScher opened the pull request:
# WPIcal

WPIcal is a tool used to “calibrate” or empirically measure the position and orientation of the Apriltags on FRC fields. This tool was inspired by team 1538, The Holy Cows, who created a command-line tool to perform this field calibration: https://github.com/TheHolyCows/cowlibration-field. WPIcal aims to streamline field calibration by combining the needed camera calibration with the field calibration process into a user-friendly application.

## Overview
WPIcal performs camera calibration using a video of either a ChArUco board (using OpenCV calibration) or a chessboard (using MRcal). Users can also upload their own camera intrinsics JSON if they so choose. The tool then measures the positions of the Apriltags on the field relative to each other. By selecting a tag to “pin” or use the default position for, WPIcal will create a field map, in which the other tags on the field will be moved to their measured positions relative to the pinned tag.

## Proposal
Integrating this tool into WPILib will ensure that all teams have access to a good calibration technique for:
* Field calibration at events
* Calibration of practice fields (which are very often imperfect)
* Apriltag "sanity check" for FTAs (tag position readout is provided with both position and rotation deltas)

## Dependencies
WPIcal uses some dependencies that will need to be added to WPILib:
- [x] [mrcal-java](https://github.com/PhotonVision/mrcal-java) is a wrapper for the mrcal library
  - [x] [mrcal](https://github.com/dkogan/mrcal) is a robust camera calibration library
  - [x] [libdogleg](https://github.com/dkogan/libdogleg) is a optimization library that mrcal depends on (there may be license issues with this library)
  - [x] [SuiteSparse](https://github.com/DrTimothyAldenDavis/SuiteSparse?tab=License-1-ov-file#readme) (this library is not explicitly used by WPIcal but is a dependency of mrcal-java)
- [x] [Ceres](http://ceres-solver.org/) is a numerical optimization library used to solve for the locations of each apriltag

## Tasks
WPIcal still has a couple of things that need to happen before it is ready:
- [x] Possible rename. WPIcal was the working title and suggestions are always appreciated!
- [x] Tool icon
- [x] Tool documentation

working version of WPIcal: https://github.com/ElliotScher/wpical

rzblue commented:
Re: nlohmann json- we already have this added. It's accessible through `wpi/json.h` and uses the `wpi` namespace instead of `nlohmann`

mcm001 commented:
I'm not super proud of the native code in mrcal-java lol. But that's just because mrcal is one huge C function with a million arguments.

mcm001 commented on wpical/src/main/native/cpp/fieldcalibration.cpp:
```diff
@@ -0,0 +1,824 @@
+
+  if (!isCalibdb)
+  {
+    for (int i = 0; i < camera_matrix.rows(); i++)
+    {
+      for (int j = 0; j < camera_matrix.cols(); j++)
+      {
+        camera_matrix(i, j) =
+            json_data["camera_matrix"][(i * camera_matrix.cols()) + j];
+      }
```
Consider replacing with a specialization of frc::from_json, like here: https://github.com/wpilibsuite/allwpilib/blob/caf424e09ca379bcf3e1807765b0b6793a648fce/apriltag/src/main/native/cpp/AprilTag.cpp#L58

mcm001 commented on wpical/src/main/native/cpp/fieldcalibration.cpp:
```diff
@@ -0,0 +1,824 @@
+    ceres::CostFunction *cost_function =
+        PoseGraphError::Create(constraint.t_begin_end);
+
+    problem.AddResidualBlock(cost_function, nullptr,
+                             pose_begin_iter->second.p.data(),
+                             pose_begin_iter->second.q.coeffs().data(),
+                             pose_end_iter->second.p.data(),
+                             pose_end_iter->second.q.coeffs().data());
+
+    problem.SetManifold(pose_begin_iter->second.q.coeffs().data(),
+                        quaternion_manifold);
+    problem.SetManifold(pose_end_iter->second.q.coeffs().data(),
+                        quaternion_manifold);
```
Consider reforumulating this as a cost function on tag corner reprojection error rather than pose error. This allows you to side-step pose ambiguity introduced by solvePNP. See https://github.com/mcm001/gtsam-playground/blob/a2de73d2fdb2c4feb39ea068dcfb7d333685babc/src/gtsam_utils.cpp#L43 & https://github.com/mcm001/gtsam-playground/blob/a2de73d2fdb2c4feb39ea068dcfb7d333685babc/src/localizer.cpp#L412, but basically the formulation here is:

```
for tag in tags:
// for each tag corner (4 per tag)
for i in range(4)
// Vec3 from camera CS to tag corner point
cameraTtag_corner = worldTtag_corner_i.relativeTo(worldTcamera_i)
// Vec2 in pixels
predicted_corner_px = camera_model.project(cameraTtag_corner )
error = predicted_corner_px - measured_corner_px
```

mcm001 replied:
This could also be formulated as a factor graph using gtsam - this is a specialization of a generic structure from motion problem (https://github.com/borglab/gtsam/blob/4.2.0/examples/SFMExampleExpressions.cpp) but I think that in both cases with the same noise model used for all tag corner pixel measurements, the math will reduce to a least squares optimization problem on reprojection error and I'd expect you to get the same output from both.

mcm001 replied:
Other prior art: https://github.com/berndpfrommer/tagslam/blob/a8afd5d2bb85e8d783831af7cedc0838e0ed53d1/src/gtsam_optimizer.cpp#L205

mcm001 replied:
https://github.com/mcm001/gtsam-playground/blob/e273fc41560ee0f88897a01c3b429cbe564f1104/src/sfm_mapper/main.cpp#L119C1-L138C6 is a snippet of a working implementation of this in gtsam.

sciencewhiz commented:
Since mrcal-java is GPLv3, this tool would have to be GPLv3, vs the rest of WPIlib which is BSD.

PeterJohnson commented:
What does this use from mrcal-java?  Would the authors be willing to relicense those pieces?

ElliotScher commented:
This tool only uses the native code, it doesn't use any of the JNI stuff. Specifically, WPIcal only uses mrcal_wrapper.cpp.

mcm001 commented:
Mrcal wrapper is a crime against humanity. Y’all can have that licensed under the WRFPL lmao.

spacey-sooty commented:
It feels like Ceres serves a similar purpose to Sleipnir which we already depend on. since we already have that as a dependency could this be migrated to Sleipnir?

spacey-sooty reviewed (commented):
Shouldn't resources be marked as binaries not as generated files in git?

PeterJohnson commented:
> Shouldn't resources like images be marked as binaries not as generated files in git? If they are in fact generated they should be placed in a directory with generated in the path as such.

I don't see where they are marked as generated files in git?  .gitattributes has not been modified.  If you're talking about .styleguide, this is consistent with how all the other tools do it currently.

calcmogul commented:
> It feels like Ceres serves a similar purpose to Sleipnir which we already depend on. since we already have that as a dependency could this be migrated to Sleipnir?

I haven't tried using Sleipnir for manifold-based optimization, so I don't know how robust it is. Instantiating `Eigen::Quaternion<sleipnir::Variable>` should work at least, after I move the NumTraits specialization into Variable.hpp.

ElliotScher commented:
Would the WPIlib community prefer to use Sleipnir to Ceres in this application?

Gold856 commented:
It would be nice if we could use Sleipnir instead of Ceres since Sleipnir is already a part of the build, which means it's less work for us.

Also, Ceres pulls in abseil as a dependency, and we avoid abseil like the plague.

spacey-sooty commented:
Another thing is Sleipnir is already reasonably well known around the FRC and WPILib space making this tool easier to maintain down the line compared to using Ceres

calcmogul commented:
I'm not too concerned about familiarity, because optimization frameworks all kinda work the same way. The optimization framework facilitating a concise expression of the problem is more important to me. Ceres looks a bit more verbose than Sleipnir or CasADi, but it's not horrible.

mcm001 commented:
Worth noting that Gtsam requires no external dependencies (boost is optional, iirc)

ElliotScher commented:
I'll try porting it over to Sleipnir to remove the Ceres dependency.

calcmogul commented:
Here's some resources on Sleipnir for reference:

Tutorial and API docs: https://sleipnirgroup.github.io/Sleipnir/docs/cpp/
Ellipse2d optimization example: https://github.com/wpilibsuite/allwpilib/blob/main/wpimath/src/main/native/cpp/geometry/Ellipse2d.cpp#L27-L53
mostly trajectory optimization examples: https://github.com/SleipnirGroup/Sleipnir/tree/main/examples

Feel free to reach out if you encounter any difficulties or bugs. I wrote Sleipnir.

mcm001 commented:
For context, this is an MVP of this using gtsam + tag corners instead of solvePNP in the optimization problem. 

https://github.com/mcm001/gtsam-playground/blob/f83a1d771ae4a55a6abc3200529217ee77961ea5/src/sfm_mapper/main.cpp

calcmogul commented:
If we can get gtsam working, that would be ideal. Sleipnir isn't designed for manifolds, while ceres and gtsam are. gtsam has fewer dependencies but requires more background knowledge (e.g., what a factor graph is) to understand the API.

mcm001 commented:
I think this problem is simple enough in formulation that that shouldn't be too crazy. I still don't understand what/why tagslam does what it does with leaves of the graph and loop closures.

Gold856 commented:
We can add Ceres as a thirdparty repo and have WPICal pull it in. If that helps maintainability, I’m all for it.

spacey-sooty commented:
https://github.com/wpilibsuite/allwpilib/pull/6929 adds ceres as an upstream utils repository

mcm001 commented:
Taking over this to repost for continuity. 

https://github.com/mcm001/gtsam-playground/blob/tag-mapper-try-2/src/sfm_mapper/main.cpp 

Managed to map both sides of a simulated 2024 field with the gtsam holy-cows-but-with a better formulation thing! Total program runtime including my prints is 50ms for 50 images of 2-5 tags each on my i7-13850HX. The images were chose randomly without regard for pose ambiguity, since that can't happen with my formation and sufficient data. Real time spent (excluding problem setup) by the dogleg solver is 14 Ms for those 16 tags and 50 snapshots

ElliotScher commented:
first draft of tool documentation has been written: https://github.com/wpilibsuite/frc-docs/pull/2691

ElliotScher commented:
Update:

- Added tests on edge case user errors to prevent unexpected crashes.
- Added .ico and .icns icon files, as well as rename them for WPIcal.

Next:
- Replace MRcal chessboard camera calibration with ChArUco for consistency with OpenCV calibration.

Aside from that, WPIcal should be ready for release once the dependencies go in and get linked to the project!

Gold856 commented:
I didn't find any Glass includes, so I think we can remove that from the dependency lists.

Gold856 commented:
Looks like you accidentally force pushed your commits away. Since you updated the main branch, you should be able to reopen the PR.

Gold856 reviewed (changes requested):
Remove dependencies on libglass, libglassnt, ntcore, wpinet, and fieldImages.

Gold856 commented on wpical/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+target_link_libraries(
+    wpical
+    wpimath
+    libglassnt
+    libglass
```
Remove

Gold856 commented on wpical/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+        target_compile_options(wpical_test PRIVATE /utf-8)
+    endif()
+    target_include_directories(wpical_test PRIVATE src/main/native/cpp src/main/native/include)
+    target_link_libraries(wpical_test wpimath libglassnt libglass googletest)
```
There are no tests, so this can be removed.

Gold856 replied:
```suggestion
    target_link_libraries(wpical_test wpimath googletest)
```

Gold856 commented on wpical/publish.gradle:
```diff
@@ -0,0 +1,124 @@
+                            description("Copies the wpical executable to the outputs directory.")
+                            destinationDirectory = outputsFolder
+
+                            archiveBaseName = '_M_' + zipBaseName
```
```suggestion
                            archiveBaseName = zipBaseName
```

Gold856 commented on wpical/publish.gradle:
```diff
@@ -0,0 +1,124 @@
+                        // We are now in the binary that we want.
+                        // This is the default application path for the ZIP task.
+                        def applicationPath = binary.executable.file
+                        def icon = file("$project.projectDir/src/main/native/mac/ov.icns")
```
```suggestion
                        def icon = file("$project.projectDir/src/main/native/mac/wpical.icns")
```

Gold856 commented:
I've assembled all the dependencies that wpical should need. To reduce merge conflicts, please resolve all review comments first, then I will create a PR on your fork that you can merge that should have everything ready to go.

ElliotScher commented:
The changes have been made, but I'm still working on making a thirdparty-mrcal repository. Can it still be merged while that is being worked on?

Gold856 commented:
Yes.

Gold856 commented:
I created the PR that assembles all the dependencies: https://github.com/ElliotScher/allwpilib/pull/2
The thirdparty-mrcal repo is techinically not necessary with this PR, and I think we'd prefer a monorepo approach when possible. I think it's also easier to link, since mrcal also depends on SuiteSparse, and it's just easier to link wpical with SuiteSparse, instead of thirdparty-mrcal to SuiteSparse, then wpical to SuiteSparse through Ceres.

ElliotScher commented:
That makes sense. Thanks for helping get this over the finish line! Finals had me super distracted for a while!

Gold856 commented on wpical/build.gradle:
```diff
@@ -0,0 +1,132 @@
+                lib project: ':thirdparty:imgui_suite', library: 'imgui', linkage: 'static'
+                lib project: ':wpimath', library: 'wpimath', linkage: 'static'
+                lib project: ':apriltag', library: 'apriltag', linkage: 'static'
+                nativeUtils.useRequiredLibrary(it, 'ceres')
```
Need to add `nativeUtils.useRequiredLibrary(it, 'opencv_static')` for OpenCV.

Gold856 reviewed (changes requested):
Changes to thirdparty sources must be managed by upstream_utils. In addition, upstream-utils.yml should be adjusted to run the upstream_utils scripts to ensure everything is in sync. Lastly, there is no need to include SuiteSparse in the monorepo; that code is part of thirdparty-ceres (although it needs to be retagged).

Gold856 commented on wpical/build.gradle:
```diff
@@ -0,0 +1,133 @@
+                lib project: ':wpiutil', library: 'wpiutil', linkage: 'static'
+                lib project: ':wpigui', library: 'wpigui', linkage: 'static'
+                lib project: ':thirdparty:imgui_suite', library: 'imgui', linkage: 'static'
+                lib project: ':wpimath', library: 'wpimath', linkage: 'static'
```
thirdparty-ceres has the Eigen headers needed. Linking to wpimath will cause conflicts between its Eigen headers and the Eigen headers from thirdparty-ceres.

Gold856 commented on wpical/CMakeLists.txt:
```diff
@@ -0,0 +1,45 @@
+target_include_directories(wpical PRIVATE src/main/native/include)
+
+find_package(OpenCV REQUIRED)
+target_link_libraries(wpical wpimath apriltaglib ${OpenCV_LIBS})
```
Remove wpimath dependency.

Gold856 commented on wpical/CMakeLists.txt:
```diff
@@ -0,0 +1,45 @@
+        target_compile_options(wpical_test PRIVATE /utf-8)
+    endif()
+    target_include_directories(wpical_test PRIVATE src/main/native/cpp src/main/native/include)
+    target_link_libraries(wpical_test wpimath googletest)
```
Remove wpimath dependency.

Gold856 commented on wpical/build.gradle:
```diff
@@ -0,0 +1,132 @@
+    dependsOn generateTask
+    dependsOn generateCppVersion
+}
+
```
```suggestion

// Suppress warnings
nativeUtils.platformConfigs.each {
    if (it.name.contains('windows')) {
        it.cppCompiler.args.add("/wd4200")
        it.cppCompiler.args.add("/wd4576")
    } else if (it.name.contains('osx')) {
        it.cCompiler.args.add("-Wno-format-nonliteral")
        it.cCompiler.args.add("-Wno-unused-function")
        it.cppCompiler.args.add("-Wno-extern-c-compat")
        it.cppCompiler.args.add("-Wno-missing-braces")
        it.cppCompiler.args.remove("-pedantic")
    } else {
        it.cCompiler.args.add("-Wno-format-nonliteral")
        it.cppCompiler.args.add("-Wno-missing-field-initializers")
        it.cppCompiler.args.remove("-pedantic")
    }
}

```
Add more suppressions as needed.

Gold856 commented on wpical/build.gradle:
```diff
@@ -0,0 +1,132 @@
+                lib project: ':apriltag', library: 'apriltag', linkage: 'static'
+                nativeUtils.useRequiredLibrary(it, 'ceres')
+                nativeUtils.useRequiredLibrary(it, 'opencv_static')
+                if (it.targetPlatform.operatingSystem.isWindows()) {
```
[StackOverflow](https://stackoverflow.com/questions/77912257/error-glog-logging-h-was-not-included-correctly-in-including-glog-in-c-pr) says defining `GLOG_NO_EXPORT` should fix the glog errors, so toss in `it.cppCompiler.define 'GLOG_NO_EXPORT'`.

ElliotScher commented:
Why is cholmod.h still not being recognized?

Gold856 commented:
The cholmod header location is slightly different in the header archive. You will need to use upstream_utils to change the include from `cholmod.h` to `suitesparse/cholmod.h`.

Gold856 commented on upstream_utils/libdogleg.py:
```diff
@@ -0,0 +1,39 @@
+    ]:
+        shutil.rmtree(os.path.join(wpical, d), ignore_errors=True)
+
+    walk_cwd_and_copy_if(
+        lambda dp, f: f.endswith("dogleg.h"),
+        os.path.join(wpical, "src/main/native/thirdparty/libdogleg/include"),
+    )
+    walk_cwd_and_copy_if(
+        lambda dp, f: f.endswith("dogleg.c"),
+        os.path.join(wpical, "src/main/native/thirdparty/libdogleg/src"),
+    )
```
```suggestion
    files = walk_cwd_and_copy_if(
        lambda dp, f: f.endswith("dogleg.h"),
        os.path.join(wpical, "src/main/native/thirdparty/libdogleg/include"),
    )
    for f in files:
        with open(f) as file:
            content = file.read()
        content = content.replace(
            "#include <cholmod.h>", "#include <suitesparse/cholmod.h>"
        )
        with open(f, "w") as file:
            file.write(content)

    files = walk_cwd_and_copy_if(
        lambda dp, f: f.endswith("dogleg.c"),
        os.path.join(wpical, "src/main/native/thirdparty/libdogleg/src"),
    )
    for f in files:
        with open(f) as file:
            content = file.read()
        content = content.replace("#warning", "// #warning")
        with open(f, "w") as file:
            file.write(content)
```
This should most of the MSVC incompatibilities as well fix the include path.

Gold856 commented:
Take a look at https://github.com/wpilibsuite/allwpilib/tree/main/upstream_utils#adding-patch-to-thirdparty-library to see how to properly modify files managed by upstream_utils and convert your changes into patches.

Gold856 commented on wpical/build.gradle:
```diff
@@ -0,0 +1,153 @@
+                lib project: ':thirdparty:imgui_suite', library: 'imgui', linkage: 'static'
+                lib project: ':apriltag', library: 'apriltag', linkage: 'static'
+                nativeUtils.useRequiredLibrary(it, 'ceres')
+                nativeUtils.useRequiredLibrary(it, 'opencv_static')
```
```suggestion
                nativeUtils.useRequiredLibrary(it, 'opencv_static')
                it.cppCompiler.define 'GLOG_NO_EXPORT'
```

Gold856 commented on wpical/build.gradle:
```diff
@@ -0,0 +1,153 @@
+    if (it.name.contains('windows')) {
+        it.cppCompiler.args.add("/wd4200")
+        it.cppCompiler.args.add("/wd4576")
+        it.cppCompiler.args.add("-DGLOG_NO_EXPORT")
```
```suggestion
```

Gold856 commented on wpical/build.gradle:
```diff
@@ -0,0 +1,153 @@
+        it.cppCompiler.args.add("-Wno-extern-c-compat")
+        it.cppCompiler.args.add("-Wno-missing-braces")
+        it.cppCompiler.args.remove("-pedantic")
+        it.cppCompiler.args.add("-DGLOG_NO_EXPORT")
```
```suggestion
```

Gold856 commented on wpical/build.gradle:
```diff
@@ -0,0 +1,153 @@
+        it.cCompiler.args.add("-Wno-format-nonliteral")
+        it.cppCompiler.args.add("-Wno-missing-field-initializers")
+        it.cppCompiler.args.remove("-pedantic")
+        it.cppCompiler.args.add("-DGLOG_NO_EXPORT")
```
```suggestion
```

ElliotScher commented:
From the stacktrace, it looks like I'm not including aruco.hpp and charuco.hpp correctly. I pulled the paths from where those headers are located in the artifactory, was that correct?

Gold856 commented:
Did you update the OpenCV version? It looks like we're still on the 4th tag (instead of the 5th)

ElliotScher commented:
yeah, in Opencv.gradle, I updated

```def opencvVersion = '4.8.0-4'```
to
```def opencvVersion = '4.8.0-5'```

Is there somewhere else it needs to be updated as well?

spacey-sooty commented on shared/config.gradle:
```diff
@@ -15,7 +15,7 @@ nativeUtils {
             opencvYear = "frc2024"
             niLibVersion = "2025.0.0"
-            opencvVersion = "4.8.0-4"
+            opencvVersion = "4.8.0-5"
```
```suggestion
            opencvVersion = "4.8.0-6"
```

spacey-sooty commented on shared/opencv.gradle:
```diff
@@ -1,4 +1,4 @@
-def opencvVersion = '4.8.0-4'
+def opencvVersion = '4.8.0-5'
```
```suggestion
def opencvVersion = '4.8.0-6'
```

mcm001 commented on wpical/src/main/native/cpp/fieldcalibration.cpp:
```diff
@@ -0,0 +1,673 @@
+      {"length", 16.541},
+      {"width", 8.211}};
+
+  // for (const auto& constraint : constraints)
+  // {
+  // wpi::json output_constraint;
+
+  // output_constraint["from_id"] = constraint.id_begin;
+  // output_constraint["to_id"] = constraint.id_end;
+
+  // output_constraint["transform"]["translation"]["x"] =
+  // constraint.t_begin_end.p[0];
+  // output_constraint["transform"]["translation"]["y"] =
+  // constraint.t_begin_end.p[1];
+  // output_constraint["transform"]["translation"]["z"] =
+  // constraint.t_begin_end.p[2];
+
+  // output_constraint["transform"]["rotation"]["x"] =
+  // constraint.t_begin_end.q.x();
+  // output_constraint["transform"]["rotation"]["y"] =
+  // constraint.t_begin_end.q.y();
+  // output_constraint["transform"]["rotation"]["z"] =
+  // constraint.t_begin_end.q.z();
+  // output_constraint["transform"]["rotation"]["w"] =
+  // constraint.t_begin_end.q.w();
+
+  // output["constraints"].push_back(output_constraint);
+
+  // std::cout << constraint.id_begin << " -- " << constraint.id_end <<
+  // std::endl;
```
Dead commented-out code?

mcm001 commented on wpical/src/main/native/cpp/fieldcalibration.cpp:
```diff
@@ -0,0 +1,673 @@
+  std::map<int, wpi::json> observed_map = ideal_map;
+
+  Eigen::Matrix<double, 4, 4> correction_a;
+  correction_a << 0, 0, -1, 0, 1, 0, 0, 0, 0, -1, 0, 0, 0, 0, 0, 1;
```
Black-magic rotation matrix? Where does this come from?

ElliotScher commented:
The errors with linking opencv seem to be related to the fact that thirdparty-opencv doesn't publish the static aruco library. It does publish the shared library though. Would it be problematic to use the shared libraries for this project instead of the static libraries?

ThadHouse commented:
Needs to be static, but that means we forgot to merge the static aruco lib in. Make an issue in the opencv repo and we can add that.

ElliotScher commented:
When I tried to build and publish thirdparty-opencv locally, ```libopencv_aruco.a``` was being built, but there's a line in publish.gradle that excludes it from being published:

```
from(resolvedDir.toFile()) {
                into project.platformPath + '/static'
                include 'libopencv*.a'
                exclude 'libopencv_*.a'
}
```
Is this the aruco lib that we need? Can we just include it in the publish step?

ThadHouse commented:
There’s a step that merges all of the static libraries. So we’d want to add that library to the merge. 

The reason we can’t do separate libs is we can’t control the link order between libraries in the same archive.

ElliotScher commented:
Could the errors with the ceres dependency have to do with the link order of the libs? They're all separate, unlike the static libs with opencv.

ThadHouse commented:
Oh. Yeah very likely.

Gold856 commented:
https://github.com/ElliotScher/allwpilib/pull/4 should fix a bunch of MSVC errors. Pretty much the only things left should be replacing all the VLAs and fixing linking.

ElliotScher commented on wpical/src/main/native/thirdparty/mrcal_java/src/mrcal_wrapper.cpp:
```diff
@@ -0,0 +1,597 @@
+  // call optimize
+
+  // Residuals
+  double c_b_packed_final[Nstate];
+  double c_x_final[Nmeasurements];
```
Should I just retag mrcal-java in upstreamutils?

mcm001 replied:
Oh forgot ya did that. Believe so! Try it and see :)

Gold856 commented:
Something's wrong with the ICO file that causes the resource file to not compile. I don't know how it was made, but the way all the current icons were made in #7116 was by opening the largest PNG in GIMP, putting in each of the smaller PNGs as their own layer, using Export As to export as an ICO file (using the .ico file extension), and selecting PNG compression for all sizes.

The icon also seems to have a lot of aliasing around edges and the thin black outline doesn't look great on smaller resolutions, so if the outline could be changed, maybe with some antialiasing, that would be great!

ElliotScher commented:
Sounds good, I'll take a stab at the icon. It seems there are still linker errors on opencv and ceres. Any ideas why that may be?

Gold856 commented:
#7476 will cover the OpenCV part. I found out that the static archive was missing objdetect, so I got a new version tagged with that fixed. For ceres, there's your PR to merge static libraries, but also my PR to build statically for Windows (we weren't before.) Once those two PRs are merged, all the linking errors _should_ be fixed.

Gold856 commented:
To get gfortran linking, you'll need to run `sudo apt-get update && sudo apt-get install -y gfortran` in the same action as the Gradle build. This actually installs gfortran into the Docker container, which allows the Gradle build to use its libraries for linking. After you do that, you can add `-lgfortran` to the linker args like normal and it should link. See https://github.com/Gold856/allwpilib/actions/runs/12168605292/job/33939749106#step:6:5.

ElliotScher commented:
Doing that would mean that linux users would need to install the gfortran package independent of wpilib thought right?

Gold856 commented:
If it's statically linked, no. (It should be statically linked)

Gold856 commented:
Here's a screenshot of me running wpical without gfortran installed in WSL:
![image](https://github.com/user-attachments/assets/2717df66-2b86-4fa0-8416-207672d988fe)

Gold856 commented on .github/workflows/gradle.yml:
```diff
@@ -48,7 +48,7 @@ jobs:
           image: ${{ matrix.container }}
           options: -v ${{ github.workspace }}:/work -w /work -e ARTIFACTORY_PUBLISH_USERNAME -e ARTIFACTORY_PUBLISH_PASSWORD -e GITHUB_REF -e CI
-          run: df . && rm -f semicolon_delimited_script && echo $GITHUB_REF && ./gradlew build --build-cache -PbuildServer -PskipJavaFormat ${{ matrix.build-options }} ${{ env.EXTRA_GRADLE_ARGS }}
+          run: sudo apt-get update && sudo apt-get install -y gfortran && ls /usr/lib/gcc/x86_64-linux-gnu/11/ && df . && rm -f semicolon_delimited_script && echo $GITHUB_REF && ls /usr/lib/gcc/x86_64-linux-gnu/11/ && ./gradlew build --build-cache -i -PbuildServer -PskipJavaFormat ${{ matrix.build-options }} ${{ env.EXTRA_GRADLE_ARGS }}
```
```suggestion
          run: sudo apt-get update && sudo apt-get install -y gfortran && df . && rm -f semicolon_delimited_script && echo $GITHUB_REF && ./gradlew build --build-cache -PbuildServer -PskipJavaFormat ${{ matrix.build-options }} ${{ env.EXTRA_GRADLE_ARGS }}
```

mcm001 commented:
> Here's a screenshot of me running wpical without gfortran installed in WSL: ![image](https://private-user-images.githubusercontent.com/117957790/392641799-2717df66-2b86-4fa0-8416-207672d988fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMzNzU0MTcsIm5iZiI6MTczMzM3NTExNywicGF0aCI6Ii8xMTc5NTc3OTAvMzkyNjQxNzk5LTI3MTdkZjY2LTJiODYtNGZhMC04NDE2LTIwNzY3MmQ5ODhmZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwNVQwNTA1MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05OGNmOTQ2MzdjNjdkOGVkMTMyMGJhZGExOTI5NjkxNDFlNTE3N2E3Mjk2NGRiYzAxODVkNTVhMGU4MTI1MmYzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Pkq1ikCExPIuBDKWJ7y-8GDvYx3i82JRemfy5gfXLnE)

you can also check via ldd if paranoid.

Gold856 commented:
/format

Gold856 commented on wpical/src/test/native/cpp/test_result_is_exact.cpp:
```diff
@@ -0,0 +1,166 @@
+#include <cstdio>
+#include <fstream>
+#include <iostream>
+#include <string>
```
```suggestion
#include <map>
#include <string>
```

spacey-sooty commented on CMakeLists.txt:
```diff
@@ -353,6 +353,8 @@ endif()
     add_subdirectory(simulation)
 endif()
+    
+add_subdirectory(wpical)
```
Should be moved within the if(WITH_GUI) statement on line 310

Gold856 commented:
/format

Gold856 commented:
Do you have H264 variants of the calibration videos? I don't think the GitHub Actions runner have an HEVC codec, so the tests fail because there's nothing to decode the HEVC video with.

mcm001 commented:
That's not why the Linux tuner is currently failing it looks like? But yeah.

mcm001 commented:
The test videos are also a little quirky. It looks like the camera hardly moves at all during calibration? For both intrinsic and field map. Maybe we should just shove them through Handbrake and transcode to something normal like h264?

ElliotScher commented:
So we need to convert the video assets from .mp4 file format into h246?

calcmogul commented:
.mp4 is a video container format that supports various codecs. They're suggesting reencoding the .mp4 to use h264 instead of HEVC.

ElliotScher commented:
@Gold856 @mcm001 Do the tests pass for you locally? I transcoded the videos using handbrake with the H.264(x264) option, but the tests still fail both on github actions and locally.

Gold856 commented:
Yeah, the tests pass locally for me but fail in CI. Basic internet research seems to indicate the audio stream might be a problem? Try taking that out as well as doing the H.264 transcode.

ElliotScher commented:
I took out the audio and it didn't work. Weren't the tests passing earlier last week?

Gold856 commented:
I did a H.264 transcode and removed the audio track with FFmpeg, and the Windows tests pass now. PR will come soon. It cleans up the rest of the code so the only failing check are Linux tests, which I determined to be due to a lack of video decode backend in OpenCV.

ElliotScher commented:
@Gold856 have you tried building WPIcal with MJPEG test assets in WSL? I tried locally on linux and it didn't seem to work properly.

Gold856 commented:
I am in the middle of doing so. OpenCV’s default capture backend for MJPEG (CV_IMAGES) thinks that the video fps is 1 for some reason, which is a problem since the detection fps is much higher. I’m not exactly sure how to resolve that, but I’m currently looking into a different capture backend.

Could you tell me more about the reasoning and logic behind the detection fps? I don’t understand why it’s needed, and I’m curious to see if we can change some behavior there.

ElliotScher commented:
I honestly can't remember why I added that. It may have been to decrease the processing time for longer videos, but I can't really remember. It can probably be removed in favor of just using the FPS of the video.

Gold856 commented:
Well, ElliotScher#6 is up, and it fixes most of the failing checks except for Linux. Let's merge that in and remove detection FPS in favor of just using the video FPS. From what I can tell, changing the detection FPS doesn't make a difference unless it's a different factor of the video FPS, which feels a little weird.

Gold856 commented:
You can format your code with wpiformat by installing it through pip. `pip install wpiformat` and then `wpiformat -f ./path/to/wpical/dirs_or_files` should do it (you may need to prepend `python -m` before wpiformat).

ElliotScher commented:
yeah, I meant to do that before I pushed, sorry...

Gold856 commented:
ElliotScher#7 fixes the Linux tests to actually process the whole video and fixes the frame number not incrementing when printed out.

If you don't have anything else to change, I think this can be taken out of draft and reviewed until it's approved for merge.

ElliotScher commented:
Awesome, I think there is one last thing. I'm getting linux opencv errors related to gtk:

![image](https://github.com/user-attachments/assets/4a090756-7027-4dec-b054-2130bbabc2b7)

do we need to enable gtk in the opencv build?

ElliotScher commented:
We could also just disable the debug windows if there are other issues with gtk.

Gold856 commented:
GTK is LGPL and we don’t want that in our OpenCV repo, so we’ll just have to turn off debug windows on Linux.

PeterJohnson commented on .github/workflows/gradle.yml:
```diff
@@ -48,7 +48,7 @@ jobs:
           image: ${{ matrix.container }}
           options: -v ${{ github.workspace }}:/work -w /work -e ARTIFACTORY_PUBLISH_USERNAME -e ARTIFACTORY_PUBLISH_PASSWORD -e GITHUB_REF -e CI
-          run: df . && rm -f semicolon_delimited_script && echo $GITHUB_REF && ./gradlew build --build-cache -PbuildServer -PskipJavaFormat ${{ matrix.build-options }} ${{ env.EXTRA_GRADLE_ARGS }}
+          run: sudo apt-get update && sudo apt-get install -y gfortran && df . && rm -f semicolon_delimited_script && echo $GITHUB_REF && ./gradlew build --build-cache -PbuildServer -PskipJavaFormat ${{ matrix.build-options }} ${{ env.EXTRA_GRADLE_ARGS }}
```
Where is this used in the build?  It's odd that it's only added in Gradle and not in the cmake builds?  If it's needed, the build system should be checking for it; this should go into the docker image (see the docker-images repo) and added to the README as a build dependency.

Gold856 replied:
A static libgfortran is needed because of the dependency on LAPACK/BLAS.

ElliotScher replied:
Should it still be moved into the docker image?

PeterJohnson replied:
If it's required, it needs to be added to the docker image and the README.

ElliotScher replied:
gfortran should be added as a build dependency in the docker-images README? or the wpical readme?

PeterJohnson replied:
It needs to be added to the docker image itself, and the allwpilib README's requirements section for Linux (so people know what is required to build allwpilib).

PeterJohnson commented on .github/workflows/comment-command.yml:
```diff
@@ -87,8 +87,8 @@ jobs:
-      - name: Install protobuf dependencies
-        run: sudo apt-get update && sudo apt-get install -y protobuf-compiler && wget https://github.com/HebiRobotics/QuickBuffers/releases/download/1.3.3/protoc-gen-quickbuf-1.3.3-linux-x86_64.exe && chmod +x protoc-gen-quickbuf-1.3.3-linux-x86_64.exe
+      - name: Install protobuf and perl dependencies
+        run: sudo apt-get update && sudo apt-get install -y protobuf-compiler liblist-moreutils-perl && wget https://github.com/HebiRobotics/QuickBuffers/releases/download/1.3.3/protoc-gen-quickbuf-1.3.3-linux-x86_64.exe && chmod +x protoc-gen-quickbuf-1.3.3-linux-x86_64.exe
```
Where is this used in pregen?

PeterJohnson commented on wpical/src/main/native/cpp/fieldmap.cpp:
```diff
@@ -0,0 +1,42 @@
+    double tagYQuat = tag.value()["pose"]["rotation"]["quaternion"]["Y"];
+    double tagZQuat = tag.value()["pose"]["rotation"]["quaternion"]["Z"];
+
+    tagVec.push_back(tag::pose(tagXPos, tagYPos, tagZPos, tagWQuat, tagXQuat,
+                               tagYQuat, tagZQuat));
```
```suggestion
    tagVec.emplace_back(tagXPos, tagYPos, tagZPos, tagWQuat, tagXQuat,
                        tagYQuat, tagZQuat);
```

PeterJohnson commented on wpical/src/main/native/cpp/fmap.cpp:
```diff
@@ -0,0 +1,44 @@
+
+#include <string>
+
+std::string fmap::singleTag(int tag, tag::pose tagpose, bool endTag) {
```
```suggestion
std::string fmap::singleTag(int tag, const tag::pose& tagpose, bool endTag) {
```

mcm001 replied:
Also. Is the all lowercase pose struct name in line with our coding standards?

PeterJohnson replied:
No, coding guidelines are for all classes and structs to use PascalCase.

PeterJohnson commented on wpical/src/main/native/cpp/fmap.cpp:
```diff
@@ -0,0 +1,44 @@
+  return fmapTag;
+}
+
+std::string fmap::convertfmap(wpi::json json) {
```
```suggestion
std::string fmap::convertfmap(const wpi::json& json) {
```

PeterJohnson commented on wpical/src/main/native/cpp/fieldmap.cpp:
```diff
@@ -0,0 +1,42 @@
+
+fieldmap::fieldmap() {}
+
+fieldmap::fieldmap(wpi::json json) {
```
```suggestion
fieldmap::fieldmap(const wpi::json& json) {
```

This should take the json by const ref.  Changing this does mean the code below changes, though, and you'll have to use the potentially-throwing const getters and catch those.

PeterJohnson commented on wpical/src/main/native/cpp/fieldmap.cpp:
```diff
@@ -0,0 +1,42 @@
+  }
+}
+
+tag::pose fieldmap::getTag(int tag) {
```
```suggestion
const tag::pose& fieldmap::getTag(size_t tag) const {
```

PeterJohnson commented on wpical/src/main/native/cpp/fieldmap.cpp:
```diff
@@ -0,0 +1,42 @@
+  return tagVec[tag - 1];
+}
+
+int fieldmap::getNumTags() {
```
```suggestion
int fieldmap::getNumTags() const {
```

PeterJohnson commented on wpical/src/main/native/cpp/fieldmap.cpp:
```diff
@@ -0,0 +1,42 @@
+  return tagVec.size();
+}
+
+double fieldmap::minimizeAngle(double angle) {
```
This seems like a utility function?  We have a function in wpimath for this.

ElliotScher replied:
linking wpimath with wpical seems to introduce issues with eigen, as we are using eigen from the ceres dependency. Am I going about this incorrectly?

Gold856 replied:
No, wpimath and ceres use totally different versions of Eigen. The apriltag library actually links to wpimath, but Gradle doesn't do transitive dependencies, and so we can kinda get away with it. The right way to do this is to make an overlay port in thirdparty-ceres that uses the same Eigen version we use, then we should be able to link everything without problems.

ElliotScher replied:
For now, is it worth removing the angle wrapping to get this over the finish line?

Gold856 replied:
I think we can fix it later.

PeterJohnson replied:
Ok.  This should at least be made a static function though.

PeterJohnson commented on wpical/src/main/native/cpp/fmap.cpp:
```diff
@@ -0,0 +1,44 @@
+  if (!endTag) {
+    fmapTag += ",";
+  }
+  return fmapTag;
```
Manually generating json like this isn't great.  This should construct a `wpi::json` and dump() it to a string--or better, return it as a wpi::json, and have the caller stringize it, since the opposite conversion is implemented directly with a json object.

PeterJohnson commented on wpical/src/main/native/cpp/fmap.cpp:
```diff
@@ -0,0 +1,44 @@
+                           i == fieldmap.getNumTags() - 1 ? true : false);
+  }
+
+  return fmapstart.append(fmapend);
```
Ditto re: returning a wpi::json instead of building a json string.

PeterJohnson commented on wpical/src/main/native/cpp/tagpose.cpp:
```diff
@@ -0,0 +1,29 @@
+  rotationMatrix = quaternion.toRotationMatrix();
+  transformMatrixFmap.setZero();
+  transformMatrixFmap.block<3, 3>(0, 0) = rotationMatrix;
+  transformMatrixFmap(0, 3) = xpos - (16.541 / 2.0);
+  transformMatrixFmap(1, 3) = ypos - (8.211 / 2.0);
```
These look like magic constants--what do they represent?  Should they actually be constant, or passed in from somewhere else?

PeterJohnson commented on wpical/src/main/native/include/fieldmap.h:
```diff
@@ -0,0 +1,23 @@
+  explicit fieldmap(wpi::json map);
+  tag::pose getTag(int tag);
+  int getNumTags();
+  double minimizeAngle(double angle);
```
See comments in fieldmap.cpp.  fieldmap(), getTag(), and getNumTags() can just be inlined here.

ElliotScher commented:
Removed gfortran from CI workflow and opened https://github.com/wpilibsuite/docker-images/pull/38. Linux gradle build fails for now.

PeterJohnson commented on wpical/src/main/native/cpp/fmap.cpp:
```diff
@@ -22,23 +22,22 @@ std::string fmap::singleTag(int tag, tag::pose tagpose, bool endTag) {
-    fmapTag += ",";
-  }
-  return fmapTag;
+  return wpi::json::parse(fmapTag);
```
Don't parse it from a built string.. you can make the json as an object directly, e.g.:
```
return {{"family", "apriltag3_36h11_classic"},
          {"id", tag},
          {"size", 165.1},
          {"transform", std::move(transformjson)},
          {"unique", true}};
```
And you can dynamically build transformjson as a json array.

PeterJohnson commented on wpical/src/main/native/cpp/tagpose.cpp:
```diff
@@ -5,6 +5,8 @@
 
 namespace tag {
+const double fieldLength = 16.541;
+const double fieldWidth = 8.211;
```
Is this truly fixed?  Doesn't it change yearly?

ElliotScher replied:
I wasn't aware of that. These values are pulled from the CAD of the field. I was under the impression that the field borders don't change.

mcm001 replied:
Consider making these units instead of magic doubles, too.

ElliotScher replied:
We would need to link wpimath for the units library right? @PeterJohnson would you prefer that the field dimensions be passed by the user side?

sciencewhiz replied:
It's approximately 54 x 27 ft, but changes a little year to year. You can look at https://github.com/wpilibsuite/allwpilib/tree/main/fieldImages/src/main/native/resources/edu/wpi/first/fields for a history

ElliotScher replied:
That's interesting. In that case, it may be worth having a field for the user to pass the field dimensions as parameters, perhaps with a default value of the 2025 field when dimensions are released on kickoff. Would that work?

ElliotScher replied:
Ahh, I forgot, the field length and width are included in the .json uploaded to wpical as the initial guess tag poses. We can just use the value from there.

Gold856 commented:
Can you click the resolve button on the review comments that have been resolved? Hard to tell what is actually resolved.

mcm001 commented on wpical/src/main/native/cpp/cameracalibration.cpp:
```diff
@@ -0,0 +1,426 @@
+#include <string>
+#include <vector>
+
+bool filter(std::vector<cv::Point2f> charuco_corners,
```
nit: should be static

mcm001 commented on wpical/src/main/native/include/fieldmap.h:
```diff
@@ -0,0 +1,60 @@
+#include <tagpose.h>
+#include <wpi/json.h>
+
+class fieldmap {
```
Class names should be pascal case

mcm001 commented on wpical/src/main/native/include/fieldmap.h:
```diff
@@ -0,0 +1,60 @@
+
+  int getNumTags() const { return tagVec.size(); }
+
+  static double minimizeAngle(double angle) {
```
what happens if I have an angle > 720? this function wont work right?

ElliotScher replied:
I realized I used radians in the commit message, but this should work better right?

mcm001 commented on wpical/src/main/native/include/fieldcalibration.h:
```diff
@@ -0,0 +1,68 @@
+};
+
+namespace fieldcalibration {
+std::tuple<Eigen::Matrix<double, 3, 3>, Eigen::Matrix<double, 8, 1>>
```
Nit for the future: I'd consider using a struct over tuples for readability. I don't see any doxygen comments to tell me what the tuple members are supposed to represent, at least I can infer from field names.

mcm001 commented on wpical/src/main/native/include/fieldcalibration.h:
```diff
@@ -0,0 +1,68 @@
+                   const Eigen::Matrix<double, 3, 3>& camera_matrix,
+                   const Eigen::Matrix<double, 8, 1>& camera_distortion,
+                   double tag_size);
+bool process_video_file(
```
Is process_video_file intended to be part of your external API surface? I only see it being used in fieldcalibration.cpp. If not, consider making static.

ElliotScher replied:
All of the functions except calibrate are only ever used in fieldcalibration. Should they all be implemented statically in fieldcalibration.h?

mcm001 replied:
If a function doesn't need to be accessible outside of a single .cpp file, they should be declared as static functions in that cpp file.

mcm001 commented on wpical/src/main/native/include/fieldcalibration.h:
```diff
@@ -36,24 +36,28 @@ struct Constraint {
-                   const Eigen::Matrix<double, 8, 1>& camera_distortion,
-                   double tag_size);
-bool process_video_file(
+static void draw_tag_cube(cv::Mat& frame,
```
If you aren't going to use draw_tag_cube outside fieldcalibration.cpp, delete it entirely from the header and move either the prototype to the top of the cpp file, or just put the whole function in the cpp file and make sure it's static. Either way, doesnt need to be here.

ElliotScher replied:
Do any of the helper functions need to be defined in the header if they're only ever being used in fieldcalibration.cpp?

spacey-sooty replied:
If they are only used in fieldcalibration.cpp then put them at the top of that file and mark them as inline

mcm001 replied:
Function prototypes only need to be in the header if you want to use the function or struct in another C++ file.

spacey-sooty commented on wpical/src/main/native/cpp/fieldcalibration.cpp:
```diff
@@ -0,0 +1,588 @@
+  const Pose m_t_ab_observed;
+};
+
+static CameraModel load_camera_model(std::string path) {
```
```suggestion
inline CameraModel load_camera_model(std::string path) {
```

mcm001 replied:
I mean, does it really matter? I feel like the compiler will sort it out

mcm001 replied:
Any particular reason for inline over static?

calcmogul replied:
Seems like a toss-up: https://stackoverflow.com/a/7767858

By the way, this function should take std::string_view instead of std::string to avoid a copy.

mcm001 replied:
any particular reason for inline over static?

spacey-sooty commented on wpical/src/main/native/cpp/fieldcalibration.cpp:
```diff
@@ -0,0 +1,588 @@
+  return camera_model;
+}
+
+static CameraModel load_camera_model(wpi::json json_data) {
```
```suggestion
inline CameraModel load_camera_model(wpi::json json_data) {
```

spacey-sooty commented on wpical/src/main/native/cpp/fieldcalibration.cpp:
```diff
@@ -0,0 +1,588 @@
+  return camera_model;
+}
+
+static std::map<int, wpi::json> load_ideal_map(std::string path) {
```
```suggestion
inline std::map<int, wpi::json> load_ideal_map(std::string path) {
```

Gold856 commented on wpical/src/main/native/include/cameracalibration.h:
```diff
@@ -0,0 +1,29 @@
+#include <opencv2/opencv.hpp>
+#include <opencv2/videoio.hpp>
+#include <wpi/json.h>
+
```
Move all headers except `<string>` to cameracalibration.cpp

Gold856 commented on wpical/src/main/native/include/fieldcalibration.h:
```diff
@@ -0,0 +1,30 @@
+
+#pragma once
+
+#include <functional>
+#include <map>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+#include <ceres/ceres.h>
+#include <opencv2/calib3d.hpp>
+#include <opencv2/core/eigen.hpp>
+#include <opencv2/core/utils/logger.hpp>
+#include <opencv2/highgui.hpp>
+#include <opencv2/imgproc.hpp>
+#include <opencv2/videoio.hpp>
+#include <wpi/json.h>
+
+#include "apriltag.h"
```
Move everything except `<string>` to fieldcalibration.cpp

Gold856 commented on wpical/src/main/native/include/fieldmap.h:
```diff
@@ -0,0 +1,61 @@
+
+#pragma once
+
+#include <numbers>
+#include <vector>
+
+#include <Eigen/Dense>
+#include <Eigen/Geometry>
+#include <tagpose.h>
+#include <wpi/json.h>
+
```
The only headers used here are `<vector>`, `<tagpose.h>`, and `<wpi/json.h>`. `<numbers>` can be swapped with `<cmath>` for `std::fmod`. Everything else can be removed.

Gold856 commented on wpical/src/main/native/include/fmap.h:
```diff
@@ -0,0 +1,16 @@
+
+#include <fieldmap.h>
+
+#include <opencv2/opencv.hpp>
```
Remove.

Gold856 commented on wpical/src/main/native/include/tagpose.h:
```diff
@@ -0,0 +1,20 @@
+
+#pragma once
+
+#include <Eigen/Dense>
```
Replace with `<Eigen/Core>`

Gold856 reviewed (changes requested):
The headers violate include what you use; in general, you only want to include what is actually used in a given file.

Also, when we were merging PRs, the .ico file got changed to Glass's .ico file; that should be fixed to have wpical's icon again.

Gold856 commented on wpical/build.gradle:
```diff
@@ -0,0 +1,249 @@
+                }
+                lib project: ':wpiutil', library: 'wpiutil', linkage: 'static'
+                lib project: ':wpigui', library: 'wpigui', linkage: 'static'
+                lib project: ':thirdparty:imgui_suite', library: 'imgui', linkage: 'static'
```
```suggestion
                lib project: ':thirdparty:imgui_suite', library: 'imguiSuite', linkage: 'static'
```

Gold856 replied:
Still need this one

Gold856 commented on wpical/src/main/native/cpp/WPIcal.cpp:
```diff
@@ -0,0 +1,588 @@
+
+#ifndef RUNNING_WPICAL_TESTS
+#ifdef _WIN32
+int __stdcall WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance,
+                      LPSTR pCmdLine, int nCmdShow) {
+  // AllocConsole();
+  // std::FILE *f;
+  // freopen_s(&f, "CONOUT$", "w", stdout);
+  // freopen_s(&f, "CONOUT$", "w", stderr);
+  // freopen_s(&f, "CONIN$", "r", stdin);
```
```suggestion
int __stdcall WinMain(void* hInstance, void* hPrevInstance, char* pCmdLine,
                      int nCmdShow) {
```

Gold856 commented:
For the record, I'm fairly sure the reason for the sudden failure is because one of the apriltag headers pulled in Windows.h. Now that the apriltag header is only in the C++ files, not the headers, the main WPICal file suddenly doesn't have Windows.h anymore, which changes how WinMain is defined.

Gold856 commented:
https://github.com/ElliotScher/allwpilib/pull/12 will fix the ICO file.

mcm001 commented on wpical/src/main/native/cpp/fieldcalibration.cpp:
```diff
@@ -0,0 +1,600 @@
+
+    bool success =
+        process_video_file(tag_detector, camera_matrix, camera_distortion,
+                           0.1651, path, poses, constraints, show_debug_window);
```
Where does this magic 0.1651 come from? (pull into a constant and justify it)

mcm001 commented on wpical/src/main/native/cpp/fieldcalibration.cpp:
```diff
@@ -0,0 +1,600 @@
+}
+
+inline bool process_video_file(
+    apriltag_detector_t* tag_detector,
```
Nit: Now that we are in allwpilib, it should be easy to use the allwpilib AprilTag wrappers. Their API is a whole lot more sane then the umich C api, LOL. Worth considering for future.

Gold856 replied:
Yeah, it would be good to use the AprilTag wrappers, but that would require wpimath and we'd have conflicting Eigen versions, which our build can't handle (yet.)

mcm001 replied:
And that's also why we can't use AprilTagFieldLayout yet right?

Gold856 replied:
Yep.

ElliotScher commented:
/format

Gold856 commented:
/format doesn't work right now. You'll have to format manually.

PeterJohnson commented on CMakeLists.txt:
```diff
@@ -142,6 +143,7 @@ if(WITH_DOCS)
     include(AddDoxygenDocs)
     add_doxygen_docs()
 endif()
+find_package(Ceres CONFIG)
```
This should be put in a `if(WITH_WPICAL)` block and be made REQUIRED.

PeterJohnson commented on CMakeLists.txt:
```diff
@@ -314,6 +316,9 @@ if(WITH_GUI)
     add_subdirectory(glass)
     add_subdirectory(outlineviewer)
     add_subdirectory(sysid)
+    if(Ceres_FOUND AND WITH_WPICAL)
```
With the change above this can just be `if(WITH_WPICAL)`.

PeterJohnson commented on README.md:
```diff
@@ -59,6 +59,8 @@ Using Gradle makes building WPILib very straightforward. It only has a few depen
 
 On macOS ARM, run `softwareupdate --install-rosetta`. This is necessary to be able to use the macOS x86 roboRIO toolchain on ARM.
 
+On linux, run `sudo apt install gfortran`. This is necessary to be able to build WPIcal on linux platforms.
```
Is there a cmake check we can add for this?

Gold856 replied:
Build-only. libgfortran should get statically linked into the build. As for a CMake check, I think specifying `Fortran` as a language in the `project` call should do it (as well as `C` and `CXX`)

ElliotScher replied:
Should this be done in the overall wpilib cmake or the wpical cmake?

Gold856 replied:
Just the wpical CMake file, because it applies only to wpical (the other projects already auto-detect the lanaguage.)

PeterJohnson commented on shared/java/javastyle.gradle:
```diff
@@ -48,7 +48,7 @@ spotless {
             include '**/*.json'
             exclude '**/build/**', '**/build-*/**', '**/bin/**'
-            exclude '**/simgui-ds.json', '**/simgui-window.json', '**/simgui.json', '**/networktables.json', '**/*test.json'
+            exclude '**/simgui-ds.json', '**/simgui-window.json', '**/simgui.json', '**/networktables.json', '**/*test.json', '**/camera calibration.json'
```
Is this file created by wpical?

ElliotScher replied:
Yes, during the tests, it runs the camera calibration which saves to a .json file.

mcm001 replied:
Is there a reason we can't use a temporary file instead of polluting the source tree?

PeterJohnson commented on .gitignore:
```diff
@@ -14,7 +14,7 @@ networktables.json
 ntcore/connectionlistenertest.json
 ntcore/timesynctest.json
-
+wpical/src/main/native/resources/camera calibration.json
```
Why are we ignoring a file in the source tree?  This doesn't smell right.  Also, in general we avoid filenames with spaces in them.

ElliotScher replied:
As the camera calibration json is created each time the tests are run, we figured it wouldn't make sense to have it saved to the repo. Should it be?

PeterJohnson replied:
Why is it get created in the src directory and not the build directory?

ElliotScher replied:
The camera calibration saves the json to the same directory as the input video it was fed. We could make the save directory user configurable which would allow for saving the camera calibration to the build directory during the tests. Would that be better?

PeterJohnson replied:
Yes, that would be better.  It’s a devops/build system smell to generate files into the source tree.

mcm001 commented:
I can't comment on binary files, but the video file `wpical/src/main/native/resources/testcalibration.avi` isn't a fantastic example of a calibration video. It's taken nearly stationary straight on to a calibration board. https://mrcal.secretsauce.net/docs-2.2/tour-choreography.html is a good resource for how to design an optimal dance

mcm001 commented:
Am i missing something stupid? I can't seem to open an AVI video that i've confirmed looks good via v4l2
```
[ INFO:0@8.614] global videoio_registry.cpp:244 VideoBackendRegistry VIDEOIO: Enabled backends(8, sorted by priority): FFMPEG(1000); GSTREAMER(990); INTEL_MFX(980); V4L2(970); CV_IMAGES(960); CV_MJPEG(950); UEYE(940); OBSENSOR(930)
[ INFO:0@8.614] global backend_plugin.cpp:369 getPluginCandidates VideoIO plugin (FFMPEG): glob is 'libopencv_videoio_ffmpeg*.so', 1 location(s)
[ INFO:0@8.614] global backend_plugin.cpp:379 getPluginCandidates     - /home/matt/Documents/GitHub/allwpilib/wpical/build/install/wpicalTest/linuxx86-64/debug/lib: 0
[ INFO:0@8.614] global backend_plugin.cpp:383 getPluginCandidates Found 0 plugin(s) for FFMPEG
[ INFO:0@8.615] global backend_plugin.cpp:369 getPluginCandidates VideoIO plugin (GSTREAMER): glob is 'libopencv_videoio_gstreamer*.so', 1 location(s)
[ INFO:0@8.615] global backend_plugin.cpp:379 getPluginCandidates     - /home/matt/Documents/GitHub/allwpilib/wpical/build/install/wpicalTest/linuxx86-64/debug/lib: 0
[ INFO:0@8.615] global backend_plugin.cpp:383 getPluginCandidates Found 0 plugin(s) for GSTREAMER
[ INFO:0@8.615] global backend_plugin.cpp:369 getPluginCandidates VideoIO plugin (INTEL_MFX): glob is 'libopencv_videoio_intel_mfx*.so', 1 location(s)
[ INFO:0@8.615] global backend_plugin.cpp:379 getPluginCandidates     - /home/matt/Documents/GitHub/allwpilib/wpical/build/install/wpicalTest/linuxx86-64/debug/lib: 0
[ INFO:0@8.615] global backend_plugin.cpp:383 getPluginCandidates Found 0 plugin(s) for INTEL_MFX
[ INFO:0@8.615] global cap_images.cpp:289 icvExtractPattern Pattern: /home/matt/Videos/wpical-testing/%04d-12-27_08-42-50.avi @ 2024
[ INFO:0@8.615] global cap_images.cpp:338 open CAP_IMAGES: File is not an image: /home/matt/Videos/wpical-testing/2024-12-27_08-42-50.avi
Unable to open video /home/matt/Videos/wpical-testing/2024-12-27_08-42-50.avi
```
![image](https://github.com/user-attachments/assets/f8bce435-4f34-4f05-ada1-f678acd4f48c)
[video.zip](https://github.com/user-attachments/files/18262323/video.zip)

ElliotScher commented:
Can you try removing the audio stream?

mcm001 commented:
Yep, removing the audio stream fixed it. FFMPEG command was `ffmpeg -i ../2024-12-27_08-42-50.mkv -f avi -c:v mjpeg -b:v 4000k -an 2024-12-27_08-42-50.avi`

ElliotScher commented:
I'll add that to the transcoding instructions for Linux

PeterJohnson commented on .gitignore:
```diff
@@ -14,7 +14,6 @@ networktables.json
 
 ntcore/connectionlistenertest.json
 ntcore/timesynctest.json
-
```
Revert this blank line removal.

Gold856 commented:
Hold on, I've got https://github.com/ElliotScher/allwpilib/pull/13 to check the minimath generated header.
